machine.go

```diff
@@ -264,10 +284,44 @@ func NewMachine(ctx context.Context, cfg Config, opts ...Opt) (*Machine, error)
 	m.machineConfig = cfg.MachineCfg
 	m.Cfg = cfg
 
+	// Use the socketPath (which must be unique to the vm) to generate a unique ID
+	// that doesn't include any path components or other special characters.
+	uniqueID, err := uniqueIDFromSocketPath(cfg.SocketPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate unique machine ID")
+	}
+	m.uniqueID = uniqueID
+
 	m.logger.Debug("Called NewMachine()")
 	return m, nil
 }
 
+func uniqueIDFromSocketPath(socketPath string) (string, error) {
+	absPath, err := filepath.Abs(socketPath)
+	if err != nil {
+		return "", errors.Wrapf(err,
+			"failed to get absolute path of socket path %q", socketPath)
+	}
+
+	// we don't assume all parent directories for the socket yet exist, so we evaluate
```

Comment:
Not so sure what you want to accomplish by doing so.

Let's say the sockerPath is `/foo/bar/baz` and `/foo` is liked to `/realfoo` which is missing, can't you check the existence of `/foo/bar/baz` directly to confirm that the path is valid?

Reply:
The problem occurs when:
1. `/foo/bar` doesn't exist yet
1. `/foo` does exist and is a symlink to `/realfoo`

In that case trying to resolve `/foo/bar` with just `EvalSymlinks` will fail with `ENOENT` whereas what we actually want is to obtain `/realfoo/bar` even though `bar` doesn't yet exist. Let me know if that makes sense

Reply:
I see. And the socket file won't be there at the time NewMachine is calling this function.